src/thor/route_action.cc

```diff
@@ -489,50 +507,87 @@ void thor_worker_t::path_arrive_by(Api& api, const std::string& costing) {
     // if we just made a leg that means we are done recording which algorithms were used
     if (path.empty())
       algorithms.clear();
-  }
 
+    return true;
+  };
+
+  auto correlated = options.locations();
+  bool allow_retry = true;
+
+  // For each pair of locations
+  auto origin = ++correlated.rbegin();
+  while (origin != correlated.rend()) {
+    auto destination = std::prev(origin);
+    if (!route_two_points(origin, destination)) {
+      // if routing failed because an intermediate waypoint was snapped to the low reachability road
+      // (such road lies in a small connectivity component that is not connected to other locations)
+      // we should leave only high reachability candidates and try to route again
+      if (allow_retry && is_through_point(*destination) &&
```

Comment:
i believe this still wont fix all cases. if you have a multipoint route that has lets say 5 locations, you could get stuck in a local maxima. imagine and island where your first 4 locations are, they all route fine but the 5th one is on the outer larger network. for htis to work you have to actually pop off all the first 3 legs you made with the first 4 low reach locations in order to find a path. though i agree solving n=1 is better than not solving it at all. i just wanted to point out that we could still have failures if we dont solve n=m.

Reply:
seems like we will find a route in your example: if we fail at 4th leg (route from 4th location that lies in a small cc to 5th location from a big cc) we will remove path edges that belong to a small cc and find some route that lies in a larger network. did I miss something ? what is the expected result here ?

Reply:
the only case that is not covered here: if we route in an island and have unreachable in one direction waypoints (like oneways that lead to deadends). here all locations will be in a small cc, so we can't retry. Such case requires additional reachability analysis: I think for now it's okay

Reply:
> did I miss something 

i think so, if you redo the 4th leg with only large components then you'll have disconnected legs. the first 3 legs will be in the small component and the last one will be out of it. im not sure what havoc that would wreak downstream

Reply:
I remove all low reachable edge candidates for all intermediate locations, not only for last one

Reply:
@kevinkreiser don't you think that we should always use `max_reachability` value from the config to set `minimun_reachabilty` filed for all locations in order to ensure
1. consistency;
2. `is_high_reach` will return true only for really big components ? because right now we use `default_reachabilty` parameter to initialize `minimum_reachability` for locations, and it equals to 50

Reply:
the way it works is that per location in the request the user can specify the min reach they want to see on a location before they consider it "highly reachable". if they dont specify this value, it gets filled out from config with a default, and yes that default is 50. in any case the location will have the number of reach which we should consider as highly reachable if an edge meets or exceeds this number. ie `location.minimum_reachability` should be the number the user or our default config thinks is highly reachable.

1. sure it would be more consistent but it would also remove the ability for the user to choose! we cant know what data the user is working with, or what their threshold is.
2. yes im saying we dont need that static high reach value that you hardcoded. the definition of high reach is already existing in the minimum_reachabiility on each location